src/Icicle/Avalanche/FromCore.hs

```diff
@@ -75,40 +80,19 @@ programFromCore namer p
   lets stmts inner
    = foldr (\(n,x) a -> Let n x a) inner stmts
 
-  -- Fold accumulator
-  initAccum (CS.SFold n ty z _) inner
-   = let write | z /= X.XValue () ty (defaultOfType ty)
-               = Write (namerAccPrefix namer n) z
-               | otherwise
-               = mempty
-
-         sub   = runIdentity $ X.transformX return (return . X.renameExp ren) inner
-         ren n'= if n == n' then namerElemPrefix namer n else n'
-
-     in write <> Read (namerElemPrefix namer n) (namerAccPrefix namer n) ty sub
-  initAccum (CS.SFilter _ ss) inner
-   = foldr initAccum inner ss
-  initAccums ss
-   = foldr initAccum mempty ss
+  createAccums _ [] inner
+   = return inner
+  createAccums subst (CS.SFold n ty z _ : ss) inner
+   = do  z' <- doSubst subst z
+         let subst' = (n,z') : subst
+         rest <- createAccums subst' ss inner
+         return $ InitAccumulator ( A.Accumulator (namerAccPrefix namer n) ty z' ) rest
+  createAccums subst (CS.SFilter _ ss : ss') inner
+   = createAccums subst (ss <> ss') inner
 
-  createAccums inner
-   = foldr createAccumMkInit inner accNames
-  createAccumMkInit (n,ty)
-   = InitAccumulator $ A.Accumulator (namerAccPrefix namer n) ty
-   $ X.XValue () ty (defaultOfType ty)
+  doSubst subs z
+   = foldM (\z' (n,x) -> X.subst () n x z') z subs
 
-  -- Nest the streams into a single loop
-  factLoopNew
-   = factLoop FactLoopNew
-
-  factLoop loopType
-   = ForeachFacts (FactBinds (C.factTimeName p) (C.factIdName p) [(C.factValName p, inputType')]) inputType' loopType
-   $ Block factStmts
-
-  inputType' = PairT (C.inputType p) TimeT
-
-  (accNames, factStmts)
-   = makeStatements p namer (C.streams p)
```

Comment:
it was sort of a combined bug in conversion to Avalanche and Dead, to do with shadowing. we decided to keep Dead simple and modify the conversion to have no shadowing.

conversion to Avalanche was producing something like

```
read x = x$acc
write x$acc = 0
read x = x$acc
write x$acc = x + 1
```

where the first "x" is dead, but Dead was confused by the shadowing.